### DISCUSSION TITLE
Conditional UI and challenges

### DISCUSSION THREAD

**Discussion by MasterKale** [MasterKale/SimpleWebAuthn#225] [Troubleshooting]
@P4sca1 asked the following in the merged browser autofill (a.k.a. conditional UI) PR:

> @MasterKale How is this supposed to work? Right now, `fetch('/generate-authentication-options')` requires the user to be signed in, so that their registered authenticators can be fetched and filled into `allowCredentials`.

For a bit more context, here's the example of how I proposed using the new capability of `startAuthentication()`:

```html
<head>
  <!-- ... -->
  <script>
    const { startAuthentication } = SimpleWebAuthnBrowser;

    fetch('/generate-authentication-options')
      .then((options) => {
        startAuthentication(options, true)
          .then(authResp => sendToServerForVerificationAndLogin)
          .catch(err => handleError);
      });
  </script>
</head>
```

Before conditional UI, challenges were associated with a username, and so successful signing of a challenge by an authenticator associated to a username meant that user was signing in. However, conditional UI represents the "usernameless" WebAuthn flow that to date hasn't received a lot of attention for all of the hubbub around "going passwordless". You can't always know the username ahead of time, so how do you handle creating a challenge to include in the response from `/generate-authentication-options`?

One answer is to create the new challenge and then temporarily store it in the server before sending it along. When the authenticator's response comes back to the server then you can check that the challenge is one you generated via some kind of session identifier, or perhaps having the front end simply return the challenge itself. So long as the challenge exists in your temporary store, then delete the challenge from the store (so it can't be reused) and perform verification on the response. If the response is legitimate, the credential ID can be used to cross reference the user that's logging in, and you can then mark them as "logged in".

(Alternatively you can reference `userHandle` for the user's ID, but since it's not signed you'd want to verify that the user handle belongs to a user that is associated with the credential ID.)

There's been some timely discussion around this over on the WebAuthn spec repo that I think will interest you: https://github.com/w3c/webauthn/issues/1764

**Comment by P4sca1** [comment]
Here's how I handle passwordless accounts for my app:
* Manually entering a username and e-mail address (or using autofill) is still required when creating a new account
* User objects in the database can now have nullable passwords
* Manually entering your username or e-mail address (or using autofill) is still required for login.

So I don't go usernameless, as it has the challenges mentioned above and my application always requires a username. For other apps, where a username (or email) is not required, it might make sense to go passwordless + usernameless for the best UX. However this would lock out all users with devices, that don't support Passkeys from your app.

When your browser does not support WebAuthn, it just displays the traditional registration / login form:
<img width="615" alt="Bildschirm­foto 2022-10-24 um 12 16 27" src="https://user-images.githubusercontent.com/10992664/197504843-76793652-f49b-4a62-8ed8-22416d71a6d5.png">
<img width="545" alt="Bildschirm­foto 2022-10-24 um 12 16 47" src="https://user-images.githubusercontent.com/10992664/197504826-7f357dd7-0384-4d6e-8d8d-871d830e2992.png">

However when WebAuthn support is detected, you have additional options:
<img width="615" alt="Bildschirm­foto 2022-10-24 um 12 16 05" src="https://user-images.githubusercontent.com/10992664/197505112-02418143-9965-4f26-9d97-29fc879cd72b.png">
Clicking "Mit Passkey registrieren" will create a passwordless account.

<img width="545" alt="Bildschirm­foto 2022-10-24 um 12 17 04" src="https://user-images.githubusercontent.com/10992664/197505128-821cfdd9-d5a6-404b-994e-401b277b2e43.png">
Cicking "Mit Passkey anmelden" will initiate an API request, searching for the user matching the given username or email and generating authentication options for that user.
This allows everyone to get the authentication options for any user. But this shouldn't be a problem, because it only includes the public keys of the authenticators assigned to the user, right? @MasterKale
